### PR TITLE
Feature/mat 2142

### PR DESCRIPTION
**Pull request by RohitKandimalla**
MAT-2142 Adding a checkbox to generate CMS/eCQM ID while creating a new Measure/Composite Measure.
## Description

- Added a Checkbox, if enabled -> will get the latest eMeasureId from MAT db and will generate a default Cql library name ex: CMS1069. eMeasure id is not saved at this point, since the user has an option to discard the creation of a new measure.
- Users can also modify the suggested (default) Cql library name provided by MAT.
- This works the same for Composite Measures as well.

image

- Label for the checkbox is changed based on the Model Type.

image

image

Edge Case 1:
- When user A starts filling this form and enabled the checkbox, which generated eMeasureId (ex:1069) and defaults Cql library name (CMS1069). but haven't clicked the save button yet.
- in the meantime, user B started filling this form and enabled the checkbox, which will generate the same eMeasureId 1069 and Cql library name CMS1069. If user B goes ahead and clicked the save button first. Then for User A, we display the following warning message.

image

Edge Case 2: 
- Consider the same scenario, where user B saved a new measure, but user A has updated his/her Cql library Name, in this case, we display the following warning message.

image

## JIRA Ticket

https://jira.cms.gov/browse/MAT-2142
## Checklist

- [x] If UI changes have been made, the google WAVE plug-in has been executed to ensure no 508 issues were introduced.
- [x] Tests are included and test edge cases.
- [x] Tests have been run locally and pass.

## Screenshots (if appropriate)
- [ ] None applicable

### CODE REVIEW

File: src/main/java/mat/client/measure/AbstractNewMeasureView.java [hunk 1]
```diff
@@ -303,6 +322,26 @@ protected void addCompositeMeasureModelType() {
...
+        generateCmsIdCheckbox.setTitle("Click to generate a CMS ID and matching Library Name upon save");
+        generateCmsIdCheckbox.setId("generateCmsId_CheckBox");
+        ((Element) generateCmsIdCheckbox.getElement().getChild(0)).setAttribute("for", "generateCmsId_Input");
+        ((Element) generateCmsIdCheckbox.getElement().getChild(0).getChild(0)).setAttribute("id", "generateCmsId_Input");
+        generateCmsIdGroup.add(generateCmsIdCheckbox);
+    }
+
+    protected void addCompositeGenerateCmsIdCheckbox() {
+        generateCmsIdCheckbox.setText("Automatically generate a eCQM ID and matching Library Name upon save");
+        generateCmsIdCheckbox.setTitle("Click to generate a eCQM ID and matching Library Name upon save");
```
Comment (jkotanchik-SB) [inline src/main/java/mat/client/measure/AbstractNewMeasureView.java:336]:
```suggestion
        generateCmsIdCheckbox.setText("Automatically generate an eCQM ID and matching Library Name upon save");
        generateCmsIdCheckbox.setTitle("Click to generate an eCQM ID and matching Library Name upon save");
```

File: src/main/java/mat/client/measure/AbstractNewMeasureView.java [hunk 2]
```diff
@@ -303,6 +322,26 @@ protected void addCompositeMeasureModelType() {
         measureModelGroup.add(buildCompositeModelTypePanel());
     }
 
+    protected void addGenerateCmsIdCheckbox() {
+        generateCmsIdCheckbox.setText("Automatically generate a CMS ID and matching Library Name upon save");
+        generateCmsIdCheckbox.setTitle("Click to generate a CMS ID and matching Library Name upon save");
```
Comment (jkotanchik-SB) [inline src/main/java/mat/client/measure/AbstractNewMeasureView.java:327]: The "upon save" part reads oddly to me, since we're generating the IDs in real-time and not when the user clicks save.

File: src/main/java/mat/client/measure/ManageMeasurePresenter.java [hunk 3]
```diff
@@ -564,16 +591,48 @@ public void onSuccess(List<? extends HasListBox> result) {
...
+        detailDisplay.getFhirModel().addValueChangeHandler(clickEvent -> {
+            if (clickEvent.getValue()) {
+                detailDisplay.getGenerateCmsIdCheckbox().setText("Automatically generate a CMS ID and matching Library Name upon save");
+                detailDisplay.getGenerateCmsIdCheckbox().setTitle("Click to generate a CMS ID and matching Library Name upon save");
+            }
+        });
+        detailDisplay.getQdmModel().addValueChangeHandler(clickEvent -> {
+            if (clickEvent.getValue()) {
+                detailDisplay.getGenerateCmsIdCheckbox().setText("Automatically generate a eCQM ID and matching Library Name upon save");
+                detailDisplay.getGenerateCmsIdCheckbox().setTitle("Automatically generate a eCQM ID and matching Library Name upon save");
```
Comment (jkotanchik-SB) [inline src/main/java/mat/client/measure/ManageMeasurePresenter.java:609]: Does the text and title need to be set twice in two separate places?

File: src/main/java/mat/client/measure/ManageMeasurePresenter.java [hunk 4]
```diff
@@ -819,6 +880,27 @@ public void onSuccess(SaveMeasureResult result) {
         }
     }
 
+    private void handleInvalidEMeasureIdGenerator(SaveMeasureResult result) {
+        InvalidEMeasureIdGeneratorDialogBox dialogBox = new InvalidEMeasureIdGeneratorDialogBox();
+        boolean isFhir = ModelTypeHelper.isFhir(currentDetails.getMeasureModel());
+        if (currentDetails.getCQLLibraryName().equalsIgnoreCase("CMS" + currentDetails.geteMeasureId())) {
+            dialogBox.displayInvalidEMeasureIdMessage((isFhir ? "CMS ID " : "eCQM ID ")
+                    + currentDetails.geteMeasureId() + " has been taken, So your ID will be " + result.geteMeasureId() + ". Do you wish to continue?");
```
Comment (jkotanchik-SB) [inline src/main/java/mat/client/measure/ManageMeasurePresenter.java:888]: Suggest dropping the ask, since there is only 1 option.
```suggestion
                    + currentDetails.geteMeasureId() + " has been taken, so your ID will be " + result.geteMeasureId() + ".");
```

File: src/main/java/mat/client/measure/ManageMeasurePresenter.java (hunk 3 above)
Comment (RohitKandimalla) [reply src/main/java/mat/client/measure/ManageMeasurePresenter.java:609]: 'AbstractNewMeasureView' sets the initial Text and Title, This handler will change text based on Measure model radio button selection.

File: src/main/java/mat/client/measure/AbstractNewMeasureView.java (hunk 2 above)
Comment (RohitKandimalla) [reply src/main/java/mat/client/measure/AbstractNewMeasureView.java:327]: https://jira.cms.gov/browse/MAT-2142
Like in the story, Do you want me to add "upon save" in parenthesis? Honestly, I missed it.

Comment (RohitKandimalla) [reply src/main/java/mat/client/measure/AbstractNewMeasureView.java]: I changed it to -> Automatically generate an eCQM ID and matching Library Name (upon save)
as suggested by Thomas.

Comment (jkotanchik-SB) [reply src/main/java/mat/client/measure/AbstractNewMeasureView.java]: That still misrepresents how the ID is generated in real time, not when saved. Hence the edge cases where, on save, there may be conflicts.

If it was generated on save, that would reduce the chance of a race condition with another user, but the user would not have the opportunity to modify the generated values.

File: src/main/java/mat/client/measure/ManageMeasurePresenter.java (hunk 3 above)
Comment (jkotanchik-SB) [reply src/main/java/mat/client/measure/ManageMeasurePresenter.java]: Gotcha, thanks for clarifying!

File: src/main/java/mat/client/measure/ManageMeasurePresenter.java [hunk 5]
```diff
@@ -819,6 +880,27 @@ public void onSuccess(SaveMeasureResult result) {
...
 
+    private void handleInvalidEMeasureIdGenerator(SaveMeasureResult result) {
+        InvalidEMeasureIdGeneratorDialogBox dialogBox = new InvalidEMeasureIdGeneratorDialogBox();
+        boolean isFhir = ModelTypeHelper.isFhir(currentDetails.getMeasureModel());
+        if (currentDetails.getCQLLibraryName().equalsIgnoreCase("CMS" + currentDetails.geteMeasureId())) {
+            dialogBox.displayInvalidEMeasureIdMessage((isFhir ? "CMS ID " : "eCQM ID ")
+                    + currentDetails.geteMeasureId() + " has been taken, so your ID will be " + result.geteMeasureId() + ".");
+        } else {
+            dialogBox.displayInvalidEMeasureIdMessage((isFhir ? "CMS ID " : "eCQM ID")
+                    + currentDetails.geteMeasureId() + " has been taken, So your ID will be " + result.geteMeasureId() + ". Please revisit "
```
Comment (jkotanchik-SB) [inline src/main/java/mat/client/measure/ManageMeasurePresenter.java:891]:
```suggestion
                    + currentDetails.geteMeasureId() + " has been taken, so the ID will be " + result.geteMeasureId() + ". Please revisit "
```

File: src/main/java/mat/client/measure/ManageMeasurePresenter.java [hunk 6]
```diff
@@ -819,6 +880,27 @@ public void onSuccess(SaveMeasureResult result) {
         }
     }
 
+    private void handleInvalidEMeasureIdGenerator(SaveMeasureResult result) {
+        InvalidEMeasureIdGeneratorDialogBox dialogBox = new InvalidEMeasureIdGeneratorDialogBox();
+        boolean isFhir = ModelTypeHelper.isFhir(currentDetails.getMeasureModel());
+        if (currentDetails.getCQLLibraryName().equalsIgnoreCase("CMS" + currentDetails.geteMeasureId())) {
+            dialogBox.displayInvalidEMeasureIdMessage((isFhir ? "CMS ID " : "eCQM ID ")
+                    + currentDetails.geteMeasureId() + " has been taken, so your ID will be " + result.geteMeasureId() + ".");
```
Comment (jkotanchik-SB) [inline src/main/java/mat/client/measure/ManageMeasurePresenter.java:888]:
```suggestion
                    + currentDetails.geteMeasureId() + " has been taken, so the ID will be " + result.geteMeasureId() + ".");
```

File: src/main/java/mat/client/measure/ManageMeasurePresenter.java [hunk 7]
```diff
@@ -2262,6 +2345,27 @@ public void onSuccess(SaveMeasureResult result) {
         });
     }
 
+    private void handleCompositeInvalidEMeasureIdGenerator(SaveMeasureResult result) {
+        InvalidEMeasureIdGeneratorDialogBox dialogBox = new InvalidEMeasureIdGeneratorDialogBox();
+        boolean isFhir = ModelTypeHelper.isFhir(currentCompositeMeasureDetails.getMeasureModel());
+        if (currentCompositeMeasureDetails.getCQLLibraryName().equalsIgnoreCase("CMS" + currentCompositeMeasureDetails.geteMeasureId())) {
+            dialogBox.displayInvalidEMeasureIdMessage((isFhir ? "CMS ID " : "eCQM ID ")
+                    + currentCompositeMeasureDetails.geteMeasureId() + " has been taken, So your ID will be " + result.geteMeasureId() + ". Do you wish to continue?");
```
Comment (jkotanchik-SB) [inline src/main/java/mat/client/measure/ManageMeasurePresenter.java:2353]:
```suggestion
                    + currentCompositeMeasureDetails.geteMeasureId() + " has been taken, so the ID will be " + result.geteMeasureId() + ". Do you wish to continue?");
```

File: src/main/java/mat/client/measure/ManageMeasurePresenter.java [hunk 8]
```diff
@@ -2262,6 +2345,27 @@ public void onSuccess(SaveMeasureResult result) {
...
+        } else {
+            dialogBox.displayInvalidEMeasureIdMessage((isFhir ? "CMS ID " : "eCQM ID")
+                    + currentCompositeMeasureDetails.geteMeasureId() + " has been taken, So your ID will be " + result.geteMeasureId() + ". Please revisit "
```
Comment (jkotanchik-SB) [inline src/main/java/mat/client/measure/ManageMeasurePresenter.java:2356]:
```suggestion
                    + currentCompositeMeasureDetails.geteMeasureId() + " has been taken, so the ID will be " + result.geteMeasureId() + ". Please revisit "
```

File: src/main/java/mat/server/service/impl/MeasurePackageServiceImpl.java [hunk 9]
```diff
@@ -351,6 +351,16 @@ public int saveAndReturnMaxEMeasureId(final Measure measure) {
         return measureDAO.saveandReturnMaxEMeasureId(measure);
     }
 
+    @Override
+    public int returnMaxEMeasureId() {
```
Comment (jkotanchik-SB) [inline src/main/java/mat/server/service/impl/MeasurePackageServiceImpl.java:355]: `MaxEMeasureId` is misleading. If it were returning the max eMeasure ID, then it should return the same value on every call, the largest possible eMeasure ID. Suggest `generateEMeasureId()` or `getNextEMeasureId()`.

Comment (RohitKandimalla) [reply src/main/java/mat/server/service/impl/MeasurePackageServiceImpl.java:355]: Yes, I will work on renaming it.

Comment (RohitKandimalla) [reply src/main/java/mat/server/service/impl/MeasurePackageServiceImpl.java]: I made changes, Can you take a look at them?

**Review by jkotanchik-SB — DISMISSED** [review]

File: src/main/java/mat/server/service/impl/MeasurePackageServiceImpl.java (hunk 9 above)
Comment (jkotanchik-SB) [reply src/main/java/mat/server/service/impl/MeasurePackageServiceImpl.java]: Looks good!

**Review by greenemcg — APPROVED** [review]
